server/sonar-server/src/main/java/org/sonar/server/es/BulkIndexer.java

```diff
@@ -192,73 +154,139 @@ public void addDeletion(IndexType indexType, String id, String routing) {
    * Note that the parameter indexName could be removed if progress logs are not needed.
    */
   public static void delete(EsClient client, String indexName, SearchRequestBuilder searchRequest) {
-    BulkIndexer bulk = new BulkIndexer(client, indexName);
+    BulkIndexer bulk = new BulkIndexer(client, indexName, Size.REGULAR);
     bulk.start();
     bulk.addDeletion(searchRequest);
     bulk.stop();
   }
 
-  @Override
-  public void stop() {
-    if (bulkRequest.numberOfActions() > 0) {
-      executeBulk();
+  private final class BulkProcessorListener implements Listener {
+
+    @Override
+    public void beforeBulk(long executionId, BulkRequest request) {
+      // no action required
     }
-    try {
-      if (semaphore.tryAcquire(concurrentRequests, 10, TimeUnit.MINUTES)) {
-        semaphore.release(concurrentRequests);
+
+    @Override
+    public void afterBulk(long executionId, BulkRequest request, BulkResponse response) {
+      counter.addAndGet(response.getItems().length);
+
+      for (BulkItemResponse item : response.getItems()) {
+        if (item.isFailed()) {
+          LOGGER.error("index [{}], type [{}], id [{}], message [{}]", item.getIndex(), item.getType(), item.getId(), item.getFailureMessage());
+        }
       }
-    } catch (InterruptedException e) {
-      throw new IllegalStateException("Elasticsearch bulk requests still being executed after 10 minutes", e);
     }
-    progress.stop();
-    client.prepareRefresh(indexName).get();
-    if (size == Size.LARGE) {
-      // optimize lucene segments and revert index settings
-      // Optimization must be done before re-applying replicas:
-      // http://www.elasticsearch.org/blog/performance-considerations-elasticsearch-indexing/
-      client.prepareForceMerge(indexName).get();
 
-      updateSettings(largeInitialSettings);
+    @Override
+    public void afterBulk(long executionId, BulkRequest req, Throwable e) {
+      LOGGER.error("Fail to execute bulk index request: " + req, e);
+    }
+  }
+
+  public enum Size {
+    /** Use this size for a limited number of documents. */
+    REGULAR,
+
+    /**
+     * Large indexing is an heavy operation that populates an index generally from scratch. Replicas and
+     * automatic refresh are disabled during bulk indexing and lucene segments are optimized at the end.
+     * Use this size for initial indexing and if you expect unusual huge numbers of documents.
+     */
+    LARGE {
+      @Override
+      SizeHandler createHandler(Runtime2 runtime2) {
+        return new LargeSizeHandler(runtime2);
+      }
+    };
+
+    SizeHandler createHandler(Runtime2 runtime2) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this unused method parameter "runtime2". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)